File: EUM.EU4Parser/ObjectExport/MissionTreeExporter.cs
using EUM.Core.Context;
using EUM.Core.Serialization;
using EUM.EU4Parser.Objects;
using EUM.Utils;
using System.Text;

namespace EUM.EU4Parser.ObjectExport
{
    public class MissionTreeExporter
    {
        internal const string MissionTreeExportFolder = "missions";

        public static void LoadAndExportMissionTrees()
        {
            var savePath = Path.Combine(ExportUtils.GetBaseModPath(), ObjectType.MissionTree.ToString());

            var missionTreeFiles = Directory.GetFiles(savePath);
            if (missionTreeFiles is null)
            {
                return;
            }

            var missionTrees = new List<MissionTree>();
            foreach (var missionTreeName in missionTreeFiles)
            {
                var missionTreePath = Path.Combine(savePath, missionTreeName);
                if (!File.Exists(missionTreePath))
                {
                    continue;
                }

                var fileContent = File.ReadAllText(missionTreePath);
                if (string.IsNullOrEmpty(fileContent))
                {
                    continue;
                }

                var missionTree = JsonSerializerExtender.Deserialize<MissionTree>(fileContent);
                if (missionTree is null)
                {
                    continue;
                }

                missionTrees.Add(missionTree);
            }

            var exportedObjects = ExportMissionTrees(missionTrees);
            ExportUtils.WriteExportedFiles(exportedObjects);
        }

        public static List<ExportedObject> ExportMissionTrees(List<MissionTree> missionTrees)
        {
            var result = new List<ExportedObject>();
            foreach (var missionTree in missionTrees)
            {
                var builder = new StringBuilder();
                var slots = missionTree.OriginalMissionSlots ?? [];
                var sortedSlots = slots
                    .Where(slot => slot is not null && slot.Number > 0 && slot.Number <= 5 && slot.Missions?.Count > 0)
                    .OrderBy(slot => slot.Number)
                    .ToList();

                foreach (var slot in sortedSlots)
                {
                    if (slot is null || slot.Missions is null)
                    {
                        continue;
                    }

                    builder.AppendLine($"{UtilTools.ToSnakeCase(slot.Name)} = {{");
                    builder.AppendLine($"{ExportUtils.GetTabs()}slot = {slot.Number}");
                    builder.AppendLine($"{ExportUtils.GetTabs()}generic = {(slot.Generic == true ? "yes" : "no")}");
                    builder.AppendLine($"{ExportUtils.GetTabs()}ai = {(slot.Ai == true ? "yes" : "no")}");
                    builder.AppendLine($"{ExportUtils.GetTabs()}has_country_shield = {(slot.HasCountryShield == true ? "yes" : "no")}");

                    if (slot.BuiltPotential?.MissionSlotPotentialEntries?.Count > 0)
                    {
                        var potential = ExportBuiltPotential(slot.BuiltPotential);
                        builder.AppendLine($"{ExportUtils.GetTabs()}potential = {{{Environment.NewLine}{potential}{Environment.NewLine}{ExportUtils.GetTabs()}}}");
                    }
                    else
                    {
                        builder.AppendLine($"{ExportUtils.GetTabs()}potential = {{ }}");
                    }

                    var missions = slot.Missions
                        .OrderBy(m => m.Position)
                        .ToList();

                    var tabCount = 2;
                    foreach (var mission in missions)
                    {
                        builder.AppendLine($"{Environment.NewLine}{ExportUtils.GetTabs()}{mission.Id ?? string.Empty} = {{");

                        builder.AppendLine($"{ExportUtils.GetTabs(tabCount)}icon = {mission.Icon}");

                        builder.AppendLine($"{ExportUtils.GetTabs(tabCount)}position = {mission.Position}");

                        var requiredMissions = mission.RequiredMissionIds ?? [];
                        if (requiredMissions.Any())
                        {
                            builder.AppendLine($"{ExportUtils.GetTabs(tabCount)}required_missions = {{");
                            builder.AppendLine($"{ExportUtils.GetTabs(tabCount + 1)}{string.Join(" ", requiredMissions)}");
                            builder.AppendLine($"{ExportUtils.GetTabs(tabCount)} }}");
                        }
                        else
                        {
                            builder.AppendLine($"{ExportUtils.GetTabs(tabCount)}required_missions = {{ }}");
                        }

                        if (mission.MissionProvincesToHighlight?.MissionProvincesToHighlightEntries?.Count > 0)
                        {
                            var provinces = ExportBuiltProvincesToHighlight(mission.MissionProvincesToHighlight);
                            builder.AppendLine($"{ExportUtils.GetTabs(tabCount)}provinces_to_highlight = {{{Environment.NewLine}{provinces}{Environment.NewLine}{ExportUtils.GetTabs(tabCount)}}}");
                        }
                        else
                        {
                            builder.AppendLine($"{ExportUtils.GetTabs(tabCount)}provinces_to_highlight = {{ }}");
                        }

                        if (mission.BuiltTrigger?.MissionTriggerEntries?.Count > 0)
                        {
                            var trigger = ExportBuiltTrigger(mission.BuiltTrigger, tabCount);
                            builder.AppendLine($"{ExportUtils.GetTabs(tabCount)}trigger = {{{Environment.NewLine}{trigger}{Environment.NewLine}{ExportUtils.GetTabs(tabCount)}}}");
                        }
                        else
                        {
                            builder.AppendLine($"{ExportUtils.GetTabs(tabCount)}trigger = {{ }}");
                        }

                        if (mission.BuiltEffect?.MissionEffectEntries?.Count > 0)
                        {
                            var effect = ExportBuiltEffect(mission.BuiltEffect, tabCount);
                            builder.AppendLine($"{ExportUtils.GetTabs(tabCount)}effect = {{{Environment.NewLine}{effect}{Environment.NewLine}{ExportUtils.GetTabs(tabCount)}}}");
                        }
                        else
                        {
                            builder.AppendLine($"{ExportUtils.GetTabs(tabCount)}effect = {{ }}");
                        }

                        builder.AppendLine($"{ExportUtils.GetTabs()}}}");
                    }

                    builder.AppendLine("}");
                }

                result.Add(new ExportedObject
                {
                    Name = missionTree.Name ?? string.Empty,
                    ObjectContent = builder.ToString(),
                    GenerateReplacementFile = missionTree.GenerateReplacementFile,
                    OriginalFileName = missionTree.OriginalFileName,
                    ObjectExportFolder = MissionTreeExportFolder,
                    ObjectType = ObjectType.MissionTree
                });
            }

            return result;
        }

        internal static string ExportBuiltEffect(MissionEffect missionEffect, int? level = null)
        {
            var result = string.Empty;
            if (missionEffect.MissionEffectEntries == null)
            {
                return result;
            }

            foreach (var entry in missionEffect.MissionEffectEntries)
            {
                if ((entry.EntryType ?? string.Empty) == "empty")
                {
                    continue;
                }

                int childLevel = level ?? 2;
                result += ExportBuiltEntryChild(entry, childLevel);
            }

            return result;
        }

        internal static string ExportBuiltProvincesToHighlight(MissionProvincesToHighlight provinces)
        {
            var result = string.Empty;
            if (provinces.MissionProvincesToHighlightEntries == null)
            {
                return result;
            }

            foreach (var entry in provinces.MissionProvincesToHighlightEntries)
            {
                if ((entry.EntryType ?? string.Empty) == "empty")
                {
                    continue;
                }

                int childLevel = 2;
                result += ExportBuiltEntryChild(entry, childLevel);
            }

            return result;
        }

        internal static string ExportBuiltTrigger(MissionTrigger trigger, int? level = null)
        {
            var result = string.Empty;
            if (trigger.MissionTriggerEntries == null)
            {
                return result;
            }

            foreach (var entry in trigger.MissionTriggerEntries)
            {
                if ((entry.EntryType ?? string.Empty) == "empty")
                {
                    continue;
                }

                int childLevel = level ?? 2;
                result += ExportBuiltEntryChild(entry, childLevel);
            }

            return result;
        }

        internal static string ExportBuiltPotential(MissionSlotPotential potential)
        {
            var result = string.Empty;
            if (potential.MissionSlotPotentialEntries == null)
            {
                return result;
            }

            foreach (var entry in potential.MissionSlotPotentialEntries)
            {
                if ((entry.EntryType ?? string.Empty) == "empty")
                {
                    continue;
                }

                int childLevel = 1;
                result += ExportBuiltEntryChild(entry, childLevel);
            }

            return result;
        }

        internal static string ExportBuiltEntryChild(TreeItemEntry entry, int level)
        {
            var result = string.Empty;
            var tabs = ExportUtils.GetTabs(level);
            var entryKey = entry.Key ?? string.Empty;

            if (entry.EntryType == "statement" && entry.Statement != null)
            {
                var statement = entry.Statement;
                result += $"{tabs}{statement.Key} = {statement.Value}{Environment.NewLine}";
            }
            else if (entry.ChildEntries != null && entry.ChildEntries.Count > 0)
            {
                if (entryKey == "effect_tooltip" || entryKey == "first_limit" || entryKey == "first_effect" ||
                    entryKey == "second_limit" || entryKey == "second_effect")
                {
                    result += $"{tabs}{entry.Key ?? string.Empty} = \"{Environment.NewLine}";
                }
                else
                {
                    result += $"{tabs}{entry.Key ?? string.Empty} = {{{Environment.NewLine}";
                }

                foreach (var childEntry in entry.ChildEntries)
                {
                    result += ExportBuiltEntryChild(childEntry, level + 1);
                }

                if (entryKey == "effect_tooltip" || entryKey == "first_limit" || entryKey == "first_effect" ||
                    entryKey == "second_limit" || entryKey == "second_effect")
                {
                    result += $"{tabs}\"{Environment.NewLine}";
                }
                else
                {
                    result += $"{tabs}}}{Environment.NewLine}";
                }
            }

            return result;
        }
    }
}

File: Photino/UserInterface/src/scripts/pdxExporters/missionTreeExporter.ts
import type { MissionEffect, MissionProvincesToHighlight, MissionSlotPotential, MissionTree, MissionTrigger } from "@/structs/missionStructs";
import { toSnakeCase } from "../utils";
import { getTabs } from "./exportUtils";
import type { TreeItemEntry } from "@/components/basic_controls/builder_tree/builderTreeItemExtender";

export function exportMissionTrees(missionTrees: MissionTree[]): string[] {
    const result: string[] = [];

    for (const missionTree of missionTrees) {
        result.push(exportMissionTree(missionTree));
    }

    return result;
}

export function exportMissionTree(missionTree: MissionTree) : string {
    const builder: string[] = [];
    const slots = missionTree.originalMissionSlots ?? [];
    const sortedSlots = slots
        .filter(slot => slot && slot.number > 0 && slot.number <= 5 && slot.missions?.length > 0)
        .sort((a, b) => a.number - b.number);

    for (const slot of sortedSlots) {
        if (!slot || !slot.missions) {
            continue;
        }

        var tabCount = 1;

        builder.push(`${toSnakeCase(slot.name)} = {`);
        builder.push(`${getTabs(tabCount)}slot = ${slot.number}`);
        builder.push(`${getTabs(tabCount)}generic = ${slot.generic ? 'yes' : 'no'}`);
        builder.push(`${getTabs(tabCount)}ai = ${slot.ai ? 'yes' : 'no'}`);
        builder.push(`${getTabs(tabCount)}has_country_shield = ${slot.hasCountryShield ? 'yes' : 'no'}`);

        if (slot.builtPotential?.missionSlotPotentialEntries?.length > 0) {
            const potential = exportBuiltPotential(slot.builtPotential, tabCount + 1);
            builder.push(`${getTabs(tabCount)}potential = {\n${potential}${getTabs(tabCount)}}`);
        } 
        else {
            builder.push(`${getTabs(tabCount)}potential = { }`);
        }

        const missions = [...slot.missions].sort((a, b) => a.position - b.position);
       
        for (const mission of missions) {
            builder.push(`\n${getTabs(tabCount)}${mission.id ?? ''} = {`);
            tabCount += 1;
            builder.push(`${getTabs(tabCount)}icon = ${mission.icon}`);
            builder.push(`${getTabs(tabCount)}position = ${mission.position}`);

            const requiredMissions = mission.requiredMissionIds ?? [];
            if (requiredMissions.length > 0) {
                builder.push(`${getTabs(tabCount)}required_missions = {`);
                builder.push(`${getTabs(tabCount + 1)}${requiredMissions.join(' ')}`);
                builder.push(`${getTabs(tabCount)}}`);
            } else {
                builder.push(`${getTabs(tabCount)}required_missions = { }`);
            }

            if (mission.missionProvincesToHighlight?.missionProvincesToHighlightEntries?.length > 0) {
                const provinces = exportBuiltProvincesToHighlight(mission.missionProvincesToHighlight, tabCount + 1);
                builder.push(`${getTabs(tabCount)}provinces_to_highlight = {\n${provinces}${getTabs(tabCount)}}`);
            } else {
                builder.push(`${getTabs(tabCount)}provinces_to_highlight = { }`);
            }

            if (mission.builtTrigger?.missionTriggerEntries?.length > 0) {
                const trigger = exportBuiltTrigger(mission.builtTrigger, tabCount + 1);
                builder.push(`${getTabs(tabCount)}trigger = {\n${trigger}${getTabs(tabCount)}}`);
            } else {
                builder.push(`${getTabs(tabCount)}trigger = { }`);
            }

            if (mission.builtEffect?.missionEffectEntries?.length > 0) {
                const effect = exportBuiltEffect(mission.builtEffect, tabCount + 1);
                builder.push(`${getTabs(tabCount)}effect = {\n${effect}${getTabs(tabCount)}}`);
            } else {
                builder.push(`${getTabs(tabCount)}effect = { }`);
            }

            tabCount -= 1;
            builder.push(`${getTabs(tabCount)}}`);
        }

        builder.push(`}\n`);
    }

    return builder.join('\n');
}

function exportBuiltEffect(effect: MissionEffect, level: number = 2): string {
    if (!effect.missionEffectEntries) return '';

    let result = '';
    for (const entry of effect.missionEffectEntries) {
        if ((entry.type ?? '') === 'empty') continue;

        result += exportBuiltEntryChild(entry, level);
    }

    return result;
}

function exportBuiltTrigger(trigger: MissionTrigger, level: number = 2): string {
    if (!trigger.missionTriggerEntries) return '';

    let result = '';
    for (const entry of trigger.missionTriggerEntries) {
        if ((entry.type ?? '') === 'empty') continue;

        result += exportBuiltEntryChild(entry, level);
    }

    return result;
}

function exportBuiltProvincesToHighlight(provinces: MissionProvincesToHighlight, level: number): string {
    if (!provinces.missionProvincesToHighlightEntries) return '';

    let result = '';
    for (const entry of provinces.missionProvincesToHighlightEntries) {
        if ((entry.type ?? '') === 'empty') continue;

        result += exportBuiltEntryChild(entry, level);
    }

    return result;
}

function exportBuiltPotential(potential: MissionSlotPotential, level: number): string {
    if (!potential.missionSlotPotentialEntries) return '';

    let result = '';
    for (const entry of potential.missionSlotPotentialEntries) {
        if ((entry.type ?? '') === 'empty') continue;

        result += exportBuiltEntryChild(entry, level);
    }

    return result;
}

function exportBuiltEntryChild(entry: TreeItemEntry, level: number): string {
    let result = '';
    const tabs = getTabs(level);
    const entryKey = entry.key ?? '';

    if (entry.type === 'statement' && entry.statement) {
        const statement = entry.statement;
        result += `${tabs}${statement.key} = ${statement.value}\n`;
    } else if (entry.childEntries && entry.childEntries.length > 0) {
        const isWrapped = ['effect_tooltip', 'first_limit', 'first_effect', 'second_limit', 'second_effect'].includes(entryKey);

        if (isWrapped) {
            result += `${tabs}${entryKey} = "\n`;
        } else {
            result += `${tabs}${entryKey} = {\n`;
        }

        for (const childEntry of entry.childEntries) {
            result += exportBuiltEntryChild(childEntry, level + 1);
        }

        if (isWrapped) {
            result += `${tabs}"\n`;
        } else {
            result += `${tabs}}\n`;
        }
    }

    return result;
}